main.go

```diff
@@ -25,23 +39,108 @@ type ReceivedMessage struct {
 	ProviderMetadata interface{}
 }
 
-// StorageBackend is used for read/write to storage
-type StorageBackend interface {
-	UploadFile(ctx context.Context, data []byte, uploadBucket string, uploadLocation string) error
-
-	ReadFile(ctx context.Context, readBucket string, readLocation string) ([]byte, error)
+//// StorageBackend is used for read/write to storage
+//type StorageBackend interface {
+//	UploadFile(ctx context.Context, data []byte, uploadBucket string, uploadLocation string) error
+//
+//	ReadFile(ctx context.Context, readBucket string, readLocation string) ([]byte, error)
+//}
+
+// StorageBackendCreator is used for read/write to storage
+type StorageBackendCreator interface {
+	CreateWriter(ctx context.Context, uploadBucket string, uploadLocation string) (io.WriteCloser, error)
+	CreateReader(ctx context.Context, uploadBucket string, uploadLocation string) (io.ReadCloser, error)
 }
 
 type Firehose struct {
-	processSettings ProcessSettings
-	storageBackend  StorageBackend
-	hedwigConsumer  *hedwig.QueueConsumer
+	processSettings       ProcessSettings
+	storageBackendCreator StorageBackendCreator
+	hedwigConsumer        *hedwig.QueueConsumer
+	hedwigFirehose        *hedwig.Firehose
+	flushLock             sync.Mutex
+	flushCh               chan error
+	messageCh             chan struct {
+		message *hedwig.Message
+		errCh   chan error
+	}
+}
+
+func (fp *Firehose) flushCron() {
+	timerCh := time.After(time.Duration(fp.processSettings.FlushAfter) * time.Second)
+	errChannelMapping := make(map[string][]chan error)
+	writerMapping := make(map[string]io.WriteCloser)
+	currentTime := time.Now()
+	// go through all msgs and write to msgtype folder
+outer:
+	for {
+		select {
+		case <-timerCh:
+			break outer
+		case messageAndChan := <-fp.messageCh:
+			message := messageAndChan.message
+			errCh := messageAndChan.errCh
+			msgType := message.Type
+			// if writer doesn't exist create in mapping
+			if _, ok := writerMapping["foo"]; !ok {
+				uploadLocation := fmt.Sprintf("%s/%s/%s", msgType, currentTime.Format("2006/1/2"), currentTime.Unix())
```

Comment:
I think you also want major version as part of this path (and thus the `writerMapping` as well). There's https://github.com/cloudchacho/hedwig-go/blob/main/types.go#L4 which might be useful.